File: backend/app/state_transitions.py
"""State transition helpers for vehicles, robots, and sessions."""

from __future__ import annotations

from app.charging import choose_target_battery, estimate_requested_energy, round_kwh
from app.config import ARRIVAL_DISTANCE_THRESHOLD, CHARGE_RATE_KW
from app.dispatch import calculate_job_priority, calculate_vehicle_priority
from app.routing import build_route_to_dock, get_vehicle_service_point, occupied_dock_ids, get_available_dock_bay, calculate_distance
from app.schemas import (
    ChargingSession,
    DispatchDecision,
    DockBay,
    ParkingSpot,
    Robot,
    RobotStatus,
    SessionStatus,
    Vehicle,
    VehiclePriority,
    VehicleStatus,
)


def request_charge(
    vehicle: Vehicle,
    sessions: list[ChargingSession],
    current_tick: int,
    energy_kwh: float | None = None,
    *,
    queue_depth: int = 0,
    manual: bool = False,
) -> tuple[Vehicle, ChargingSession] | None:
    if vehicle.status not in (VehicleStatus.parked, VehicleStatus.completed):
        return None
    if not vehicle.spot_id:
        return None

    target = choose_target_battery(vehicle, queue_depth, current_tick)
    requested = round_kwh(
        energy_kwh
        if energy_kwh is not None
        else estimate_requested_energy(vehicle, target_battery=target, manual=manual)
    )
    if requested < 5:
        return None

    priority = calculate_vehicle_priority(
        vehicle.model_copy(update={"requested_energy_kwh": requested, "target_battery": target}),
        current_tick,
    )
    score, _ = calculate_job_priority(
        ChargingSession(
            id="temp",
            vehicle_id=vehicle.id,
            spot_id=vehicle.spot_id,
            status=SessionStatus.queued,
            requested_energy_kwh=requested,
            created_tick=current_tick,
        ),
        vehicle.model_copy(update={"requested_energy_kwh": requested, "target_battery": target}),
        current_tick,
    )

    session = ChargingSession(
        id=f"S-{current_tick}-{vehicle.id[-4:]}",
        vehicle_id=vehicle.id,
        spot_id=vehicle.spot_id,
        status=SessionStatus.queued,
        requested_energy_kwh=requested,
        priority_score=score,
        created_tick=current_tick,
        charge_rate_kw=CHARGE_RATE_KW,
    )
    updated_vehicle = vehicle.model_copy(update={
        "status": VehicleStatus.waiting,
        "assigned_robot_id": None,
        "requested_energy_kwh": requested,
        "target_battery": target,
        "priority": priority,
    })
    return updated_vehicle, session


def assign_robot(
    vehicle: Vehicle,
    decision: DispatchDecision,
    robots: list[Robot],
    sessions: list[ChargingSession],
) -> tuple[list[Robot], Vehicle, list[ChargingSession]]:
    robots = [
        r.model_copy(update={
            "status": RobotStatus.en_route,
            "assigned_vehicle_id": vehicle.id,
            "assigned_session_id": decision.session_id,
            "dock_bay_id": None,
            "fault_type": None,
            "route": decision.route,
            "route_index": 0,
        }) if r.id == decision.selected_robot_id else r
        for r in robots
    ]
    vehicle = vehicle.model_copy(update={
        "assigned_robot_id": decision.selected_robot_id,
        "status": VehicleStatus.assigned,
    })
    sessions = [
        s.model_copy(update={
            "status": SessionStatus.en_route,
            "robot_id": decision.selected_robot_id,
        }) if s.vehicle_id == vehicle.id and s.status in (
            SessionStatus.queued, SessionStatus.interrupted, SessionStatus.assigned
        ) else s
        for s in sessions
    ]
    return robots, vehicle, sessions


def start_charging(
    robot_id: str,
    vehicle_id: str,
    robots: list[Robot],
    vehicles: list[Vehicle],
    sessions: list[ChargingSession],
    current_tick: int,
    spots: list[ParkingSpot] | None = None,
) -> tuple[list[Robot], list[Vehicle], list[ChargingSession]]:
    vehicle = next((v for v in vehicles if v.id == vehicle_id), None)
    robot = next((r for r in robots if r.id == robot_id), None)
    if not vehicle or not robot or robot.status == RobotStatus.faulted:
        return robots, vehicles, sessions

    session = next(
        (s for s in sessions if s.vehicle_id == vehicle_id and s.status in (
            SessionStatus.queued, SessionStatus.assigned, SessionStatus.en_route,
            SessionStatus.interrupted, SessionStatus.active,
        )),
        None,
    )
    if session and session.status == SessionStatus.active and session.robot_id == robot_id:
        return robots, vehicles, sessions
    if session and session.status == SessionStatus.completed:
        return robots, vehicles, sessions
    if vehicle.status == VehicleStatus.completed:
        return robots, vehicles, sessions

    service = robot.position
    if spots and vehicle.spot_id:
        spot = next((s for s in spots if s.id == vehicle.spot_id), None)
        if spot:
            left = get_vehicle_service_point(spot, "left")
            right = get_vehicle_service_point(spot, "right")
            # Only snap to a bay the robot is already near — never jump across the car.
            near_left = calculate_distance(robot.position, left) <= ARRIVAL_DISTANCE_THRESHOLD * 2.0
            near_right = calculate_distance(robot.position, right) <= ARRIVAL_DISTANCE_THRESHOLD * 2.0
            if near_left and (not near_right or calculate_distance(robot.position, left) <= calculate_distance(robot.position, right)):
                service = left
            elif near_right:
                service = right
            else:
                service = robot.position

    robots = [
        r.model_copy(update={
            "status": RobotStatus.charging,
            "route": [],
            "route_index": 0,
            "position": service,
            "assigned_vehicle_id": vehicle_id,
            "assigned_session_id": session.id if session else r.assigned_session_id,
            "last_yield_tick": 0,
        }) if r.id == robot_id else r
        for r in robots
    ]
    vehicles = [
        v.model_copy(update={
            "status": VehicleStatus.charging,
            "assigned_robot_id": robot_id,
        }) if v.id == vehicle_id else v
        for v in vehicles
    ]
    sessions = [
        s.model_copy(update={
            "status": SessionStatus.active,
            "robot_id": robot_id,
            "started_tick": s.started_tick if s.started_tick is not None else current_tick,
        }) if s.vehicle_id == vehicle_id and s.status in (
            SessionStatus.queued, SessionStatus.assigned, SessionStatus.en_route,
            SessionStatus.interrupted, SessionStatus.active,
        ) else s
        for s in sessions
    ]
    return robots, vehicles, sessions


def complete_charging(
    robot_id: str,
    vehicle_id: str,
    robots: list[Robot],
    vehicles: list[Vehicle],
    sessions: list[ChargingSession],
    dock_bays: list[DockBay],
    blocked_lane: bool,
    current_tick: int,
) -> tuple[list[Robot], list[Vehicle], list[ChargingSession]]:
    vehicles = [
        v.model_copy(update={
            "status": VehicleStatus.completed,
            "assigned_robot_id": None,
            "completed_at_tick": current_tick,
        }) if v.id == vehicle_id else v
        for v in vehicles
    ]
    sessions = [
        s.model_copy(update={
            "status": SessionStatus.completed,
            "delivered_energy_kwh": round_kwh(s.requested_energy_kwh),
            "completed_tick": current_tick,
        }) if s.vehicle_id == vehicle_id and s.status == SessionStatus.active else s
        for s in sessions
    ]

    claimed = occupied_dock_ids(robots, exclude_robot_id=robot_id)
    bay = get_available_dock_bay(dock_bays, claimed, robot_id=robot_id)
    robot = next((r for r in robots if r.id == robot_id), None)

    def update_robot(r: Robot) -> Robot:
        if r.id != robot_id:
            return r
        if not bay or not robot:
            return r.model_copy(update={
                "status": RobotStatus.idle,
                "assigned_vehicle_id": None,
                "dock_bay_id": None,
                "route": [],
                "route_index": 0,
            })
        route = build_route_to_dock(robot.position, bay, blocked_lane)
        return r.model_copy(update={
            "status": RobotStatus.returning,
            "assigned_vehicle_id": None,
            "dock_bay_id": bay.id,
            "route": route,
            "route_index": 0,
        })

    robots = [update_robot(r) for r in robots]
    return robots, vehicles, sessions


def dock_robot(robot_id: str, robots: list[Robot], dock_bays: list[DockBay] | None = None) -> list[Robot]:
    def update(r: Robot) -> Robot:
        if r.id != robot_id:
            return r
        bay = None
        if dock_bays and r.dock_bay_id:
            bay = next((b for b in dock_bays if b.id == r.dock_bay_id), None)
        return r.model_copy(update={
            "status": RobotStatus.docked,
            "position": bay.position if bay else r.position,
            "route": [],
            "route_index": 0,
            "assigned_vehicle_id": None,
        })
    return [update(r) for r in robots]


def requeue_vehicle(
    vehicle_id: str,
    vehicles: list[Vehicle],
    sessions: list[ChargingSession],
) -> tuple[list[Vehicle], list[ChargingSession]]:
    vehicles = [
        v.model_copy(update={"status": VehicleStatus.waiting})
        if v.id == vehicle_id else v
        for v in vehicles
    ]
    sessions = [
        s.model_copy(update={"status": SessionStatus.queued})
        if s.vehicle_id == vehicle_id and s.status == SessionStatus.interrupted else s
        for s in sessions
    ]
    return vehicles, sessions


def vehicle_parks(vehicle: Vehicle, spot: ParkingSpot) -> tuple[Vehicle, ParkingSpot]:
    heading = 0 if spot.row == "top" else 180
    parked = vehicle.model_copy(update={
        "status": VehicleStatus.parked,
        "spot_id": spot.id,
        "position": spot.position,
        "route": [],
        "route_index": 0,
        "heading": heading,
    })
    updated_spot = spot.model_copy(update={
        "occupied_vehicle_id": vehicle.id,
        "reserved_vehicle_id": None,
    })
    return parked, updated_spot


def vehicle_departs(vehicle: Vehicle, spot: ParkingSpot, exit_route: list) -> tuple[Vehicle, ParkingSpot]:
    # Keep spot_id + occupancy until the car fully exits so nothing can spawn on top.
    leaving = vehicle.model_copy(update={
        "status": VehicleStatus.leaving,
        "route": exit_route,
        "route_index": 0,
        "spot_id": spot.id,
    })
    held = spot.model_copy(update={
        "occupied_vehicle_id": vehicle.id,
        "reserved_vehicle_id": None,
    })
    return leaving, held


def clear_spot_after_departure(spot: ParkingSpot, vehicle_id: str) -> ParkingSpot:
    if spot.occupied_vehicle_id != vehicle_id and spot.reserved_vehicle_id != vehicle_id:
        return spot
    return spot.model_copy(update={
        "occupied_vehicle_id": None if spot.occupied_vehicle_id == vehicle_id else spot.occupied_vehicle_id,
        "reserved_vehicle_id": None if spot.reserved_vehicle_id == vehicle_id else spot.reserved_vehicle_id,
    })

File: frontend/src/lib/ops/stateTransitions.ts
import { advanceCharging } from "../charging";
import { formatKwh } from "../format";
import type {
  ChargingSession,
  DockBay,
  EventLogItem,
  FaultType,
  GaragePosition,
  ParkingSpot,
  Robot,
  Vehicle,
} from "../types";
import { roundKwh } from "../vehicleActions";
import type { DispatchDecision } from "../dispatch";
import { buildRouteToDock, getAvailableDockBay, getVehicleServicePoint } from "../routes";
import { CHARGE_RATE_KW } from "./constants";
import { calculateJobPriority } from "./dispatch";

export function nowLabel(): string {
  return new Date().toLocaleTimeString([], { hour: "2-digit", minute: "2-digit", hour12: false });
}

export function createEvent(
  message: string,
  type: EventLogItem["type"],
  idSuffix = "",
): EventLogItem {
  return {
    id: `E-${Date.now()}-${idSuffix || Math.random().toString(36).slice(2, 6)}`,
    message,
    timestamp: nowLabel(),
    type,
  };
}

export function requestCharge(
  vehicle: Vehicle,
  sessions: ChargingSession[],
  currentTick: number,
  energyKwh?: number,
): { vehicle: Vehicle; session: ChargingSession; event: EventLogItem } | null {
  if (vehicle.status !== "parked" && vehicle.status !== "completed") return null;
  if (!vehicle.spotId) return null;

  const requestedKwh = roundKwh(energyKwh ?? vehicle.requestedEnergyKwh ?? 22);
  const priority = calculateJobPriority(vehicle, currentTick, 0);

  const session: ChargingSession = {
    id: `S-${Date.now()}`,
    vehicleId: vehicle.id,
    spotId: vehicle.spotId,
    robotId: null,
    status: "queued",
    energyKwh: 0,
    requestedKwh,
    etaSeconds: null,
    startedAt: nowLabel(),
    priorityScore: priority.score,
    chargeRateKw: CHARGE_RATE_KW,
    createdTick: currentTick,
  };

  const updatedVehicle: Vehicle = {
    ...vehicle,
    status: "waiting",
    assignedRobotId: null,
    requestedEnergyKwh: requestedKwh,
    priority: priority.urgencyLabel,
  };

  return {
    vehicle: updatedVehicle,
    session,
    event: createEvent(
      `${vehicle.id} requested ${formatKwh(requestedKwh)} at ${vehicle.spotId}. Priority: ${priority.urgencyLabel.toLowerCase()}.`,
      "request",
    ),
  };
}

export function assignRobot(
  vehicle: Vehicle,
  decision: DispatchDecision,
  robots: Robot[],
  sessions: ChargingSession[],
  reassignment = false,
): {
  robots: Robot[];
  vehicle: Vehicle;
  sessions: ChargingSession[];
  event: EventLogItem;
} {
  const nextRobots = robots.map((robot) => (
    robot.id === decision.selectedRobotId
      ? {
          ...robot,
          status: "en-route" as const,
          assignedVehicleId: vehicle.id,
          dockBayId: null,
          faultType: null,
          route: decision.route,
          routeIndex: 0,
          targetPosition: decision.route[decision.route.length - 1] ?? null,
          motionState: "moving" as const,
        }
      : robot
  ));

  const nextVehicle: Vehicle = {
    ...vehicle,
    assignedRobotId: decision.selectedRobotId,
    status: "assigned",
  };

  const nextSessions = sessions.map((session) => (
    session.vehicleId === vehicle.id
    && (session.status === "queued" || session.status === "interrupted" || session.status === "assigned")
      ? {
          ...session,
          status: "en_route" as const,
          robotId: decision.selectedRobotId,
          etaSeconds: decision.etaSeconds,
        }
      : session
  ));

  return {
    robots: nextRobots,
    vehicle: nextVehicle,
    sessions: nextSessions,
    event: createEvent(
      reassignment
        ? `${vehicle.id} reassigned to ${decision.selectedRobotId}`
        : `${decision.selectedRobotId} dispatched to ${vehicle.id}, ETA ${Math.round(decision.etaSeconds)}s`,
      reassignment ? "reassignment" : "dispatch",
    ),
  };
}

export function startCharging(
  robotId: string,
  vehicleId: string,
  robots: Robot[],
  vehicles: Vehicle[],
  sessions: ChargingSession[],
  currentTick: number,
  spots?: ParkingSpot[],
): {
  robots: Robot[];
  vehicles: Vehicle[];
  sessions: ChargingSession[];
  event: EventLogItem;
} {
  const robot = robots.find((item) => item.id === robotId);
  const vehicle = vehicles.find((item) => item.id === vehicleId);
  if (!robot || !vehicle || robot.status === "faulted") {
    return {
      robots,
      vehicles,
      sessions,
      event: createEvent(`${robotId} could not start charging ${vehicleId}.`, "dispatch"),
    };
  }

  const session = sessions.find((item) => (
    item.vehicleId === vehicleId
    && ["queued", "assigned", "en_route", "interrupted", "active"].includes(item.status)
  ));
  if (session?.status === "active" && session.robotId === robotId) {
    return { robots, vehicles, sessions, event: createEvent(`${robotId} already charging ${vehicleId}`, "charging") };
  }
  if (session?.status === "completed" || vehicle.status === "completed") {
    return { robots, vehicles, sessions, event: createEvent(`${vehicleId} already completed.`, "charging") };
  }

  let service = robot.position;
  if (spots && vehicle.spotId) {
    const spot = spots.find((item) => item.id === vehicle.spotId);
    if (spot) service = getVehicleServicePoint(spot);
  }

  const nextRobots = robots.map((item) => (
    item.id === robotId
      ? {
          ...item,
          status: "charging" as const,
          position: { ...service },
          route: [],
          routeIndex: 0,
          motionState: "charging" as const,
          assignedVehicleId: vehicleId,
        }
      : item
  ));
  const nextVehicles = vehicles.map((item) => (
    item.id === vehicleId
      ? { ...item, status: "charging" as const, assignedRobotId: robotId }
      : item
  ));
  const nextSessions = sessions.map((item) => (
    item.vehicleId === vehicleId
    && ["queued", "assigned", "en_route", "interrupted", "active"].includes(item.status)
      ? {
          ...item,
          status: "active" as const,
          robotId,
          startedAt: item.startedAt === "—" || !item.startedAt ? nowLabel() : item.startedAt,
          etaSeconds: null,
          startedTick: item.startedTick ?? currentTick,
        }
      : item
  ));

  return {
    robots: nextRobots,
    vehicles: nextVehicles,
    sessions: nextSessions,
    event: createEvent(`${robotId} arrived at ${vehicle.spotId ?? "bay"}. Charging started.`, "charging"),
  };
}

export function completeCharging(
  robotId: string,
  vehicleId: string,
  robots: Robot[],
  vehicles: Vehicle[],
  sessions: ChargingSession[],
  dockBays: DockBay[],
  laneBlocked: boolean,
  currentTick: number,
): {
  robots: Robot[];
  vehicles: Vehicle[];
  sessions: ChargingSession[];
  event: EventLogItem;
} {
  const nextVehicles = vehicles.map((vehicle) => (
    vehicle.id === vehicleId
      ? {
          ...vehicle,
          status: "completed" as const,
          assignedRobotId: null,
          completedAtTick: currentTick,
        }
      : vehicle
  ));
  const nextSessions = sessions.map((session) => (
    session.vehicleId === vehicleId && session.status === "active"
      ? {
          ...session,
          status: "completed" as const,
          energyKwh: roundKwh(session.requestedKwh),
          etaSeconds: null,
          completedTick: currentTick,
        }
      : session
  ));

  const bay = getAvailableDockBay(robots, dockBays, robotId);
  const robot = robots.find((item) => item.id === robotId);

  const nextRobots = robots.map((item) => {
    if (item.id !== robotId) return item;
    if (!bay || !robot) {
      return {
        ...item,
        status: "idle" as const,
        assignedVehicleId: null,
        dockBayId: null,
        route: [],
        routeIndex: 0,
        targetPosition: null,
      };
    }
    const route = buildRouteToDock(robot.position, bay, { laneBlocked });
    return {
      ...item,
      status: "returning" as const,
      assignedVehicleId: null,
      dockBayId: bay.id,
      route,
      routeIndex: 0,
      targetPosition: bay.position,
      motionState: "moving" as const,
    };
  });

  return {
    robots: nextRobots,
    vehicles: nextVehicles,
    sessions: nextSessions,
    event: createEvent(`${vehicleId} charge complete. ${robotId} returning to dock.`, "charging"),
  };
}

export function dockRobot(
  robotId: string,
  robots: Robot[],
): { robots: Robot[]; event: EventLogItem } {
  const nextRobots = robots.map((robot) => {
    if (robot.id !== robotId) return robot;
    return {
      ...robot,
      status: "docked" as const,
      route: [],
      routeIndex: 0,
      assignedVehicleId: null,
      motionState: "docked" as const,
    };
  });
  return {
    robots: nextRobots,
    event: createEvent(`${robotId} docked and started recharging`, "dock"),
  };
}

export function requeueVehicle(
  vehicleId: string,
  vehicles: Vehicle[],
  sessions: ChargingSession[],
): { vehicles: Vehicle[]; sessions: ChargingSession[] } {
  const vehiclesUpdated = vehicles.map((vehicle) => (
    vehicle.id === vehicleId ? { ...vehicle, status: "waiting" as const } : vehicle
  ));
  const sessionsUpdated = sessions.map((session) => (
    session.vehicleId === vehicleId && session.status === "interrupted"
      ? { ...session, status: "queued" as const }
      : session
  ));
  return { vehicles: vehiclesUpdated, sessions: sessionsUpdated };
}

export function clearFault(
  robotId: string,
  robots: Robot[],
  dockBays: DockBay[],
  laneBlocked: boolean,
): { robots: Robot[]; event: EventLogItem } | null {
  const robot = robots.find((item) => item.id === robotId);
  if (!robot || robot.status !== "faulted") return null;

  const bay = getAvailableDockBay(robots, dockBays, robotId);

  const nextRobots = robots.map((item) => {
    if (item.id !== robotId) return item;
    if (!bay) return { ...item, status: "idle" as const, faultType: null };
    const nearDock = Math.hypot(item.position.x - bay.position.x, item.position.y - bay.position.y) < 2;
    if (nearDock) {
      return {
        ...item,
        status: "docked" as const,
        faultType: null,
        dockBayId: bay.id,
        position: { ...bay.position },
        route: [],
        routeIndex: 0,
        targetPosition: null,
        motionState: "docked" as const,
      };
    }
    const route = buildRouteToDock(item.position, bay, { laneBlocked });
    return {
      ...item,
      status: "returning" as const,
      faultType: null,
      dockBayId: bay.id,
      route,
      routeIndex: 0,
      targetPosition: bay.position,
      motionState: "moving" as const,
    };
  });

  return {
    robots: nextRobots,
    event: createEvent(
      bay
        ? `${robotId} fault cleared. Returning to dock.`
        : `${robotId} fault cleared and available.`,
      "fault",
    ),
  };
}

export function vehicleParks(
  vehicle: Vehicle,
  spot: ParkingSpot,
): { vehicle: Vehicle; spot: ParkingSpot; event: EventLogItem } {
  const parked: Vehicle = {
    ...vehicle,
    status: "parked",
    spotId: spot.id,
    position: { ...spot.position },
    route: [],
    routeIndex: 0,
    heading: spot.rotation,
  };
  const updatedSpot: ParkingSpot = {
    ...spot,
    vehicleId: vehicle.id,
    occupiedVehicleId: vehicle.id,
    reservedVehicleId: null,
  };
  return {
    vehicle: parked,
    spot: updatedSpot,
    event: createEvent(
      `${vehicle.id} parked at ${spot.id} with ${Math.round(vehicle.battery)}% battery.`,
      "arrival",
    ),
  };
}

export function vehicleDeparts(
  vehicle: Vehicle,
  spot: ParkingSpot,
  exitRoute: GaragePosition[],
): { vehicle: Vehicle; spot: ParkingSpot; event: EventLogItem } {
  const leaving: Vehicle = {
    ...vehicle,
    status: "leaving",
    route: exitRoute,
    routeIndex: 0,
    spotId: null,
  };
  const clearedSpot: ParkingSpot = {
    ...spot,
    vehicleId: null,
    occupiedVehicleId: null,
    reservedVehicleId: null,
  };
  return {
    vehicle: leaving,
    spot: clearedSpot,
    event: createEvent(`${vehicle.id} departing garage.`, "departure"),
  };
}

export function advanceChargingStep(
  vehicle: Vehicle,
  session: ChargingSession,
  elapsedSeconds: number,
): {
  vehicle: Vehicle;
  session: ChargingSession;
  deliveredKwh: number;
  complete: boolean;
} {
  return advanceCharging(vehicle, session, elapsedSeconds);
}

export function applyFaultToState(
  robot: Robot,
  faultType: FaultType,
  robots: Robot[],
  vehicles: Vehicle[],
  sessions: ChargingSession[],
): {
  robots: Robot[];
  vehicles: Vehicle[];
  sessions: ChargingSession[];
  events: EventLogItem[];
  vehicleId: string | null;
} {
  const vehicleId = robot.assignedVehicleId;
  const label = faultType.replace(/_/g, " ");
  // Faulted robots freeze where they are — never side-step / teleport.
  const nextRobots = robots.map((item) => (
    item.id === robot.id
      ? {
          ...item,
          status: "faulted" as const,
          faultType,
          assignedVehicleId: null,
          dockBayId: null,
          route: [],
          routeIndex: 0,
          targetPosition: null,
          // keep item.position unchanged
        }
      : item
  ));

  const events: EventLogItem[] = [];
  if (vehicleId) {
    events.push(createEvent(`${robot.id} connector timeout while charging ${vehicleId}.`, "fault"));
    const session = sessions.find((s) => s.vehicleId === vehicleId && (s.status === "active" || s.status === "interrupted"));
    const delivered = session ? formatKwh(session.energyKwh) : "0.0 kWh";
    events.push(createEvent(`${vehicleId} requeued with ${delivered} already delivered.`, "fault"));
  } else {
    events.push(createEvent(`${robot.id} faulted: ${label}.`, "fault"));
  }

  const nextVehicles = vehicles.map((vehicle) => (
    vehicle.id === vehicleId
      ? { ...vehicle, status: "backup-needed" as const, assignedRobotId: null }
      : vehicle
  ));

  const nextSessions = sessions.map((session) => (
    session.vehicleId === vehicleId && session.status !== "completed"
      ? { ...session, status: "interrupted" as const, robotId: null, etaSeconds: null }
      : session
  ));

  return { robots: nextRobots, vehicles: nextVehicles, sessions: nextSessions, events, vehicleId };
}

export function shouldRequestCharge(vehicle: Vehicle): boolean {
  if (vehicle.status !== "parked") return false;
  if (vehicle.requestedEnergyKwh != null && vehicle.requestedEnergyKwh > 0) return true;
  const gap = vehicle.targetBattery - vehicle.battery;
  return vehicle.battery < 45 || gap >= 20 || vehicle.priority === "Urgent";
}

export function estimateRequestedEnergy(vehicle: Vehicle): number {
  if (vehicle.requestedEnergyKwh != null && vehicle.requestedEnergyKwh > 0) {
    return roundKwh(vehicle.requestedEnergyKwh);
  }
  const gap = Math.max(0, vehicle.targetBattery - vehicle.battery);
  return roundKwh(Math.min(35, Math.max(8, (gap / 100) * 75)));
}
